app/javascript/spec/ansible-repository-form/helpers.spec.js

```diff
@@ -0,0 +1,89 @@
+
+const TESTCASES = [
+  [ // Example Case
+    { scm_url: 'git@github.com:IonicaBizau/node-git-url-parse.git' },
+    {},
+  ], [
+    { scm_url: 'https://github.com/IonicaBizau/node-git-url-parse.git' },
+    {},
+  ], [
+    { scm_url: 'https://github.com/IonicaBizau/git-url-parse/blob/master/test/index.js' },
+    {},
+  ], [
+    { scm_url: 'https://github.com/IonicaBizau/node-git-url-parse.git' },
+    {},
+  ], [
+    { scm_url: 'git@github.com:IonicaBizau/node-git-url-parse.git' },
+    {},
+  ], [
+    { scm_url: '' },
+    {},
+  ], [
+    {}, // url not set
+    {},
+  ], [
+    { scm_url: '_' },
+    {},
+  ], [
+    { scm_url: 'scheme' },
+    {},
+  ], [
+    { scm_url: 'scheme:' },
+    {},
+  ], [
+    { scm_url: 'scheme:/' },
+    {},
+  ], [
+    { scm_url: 'scheme://' },
+    { scm_url: __('URL must include a protocol (http://, https:// or file://) or be a valid SSH path (user@server:path or ssh://user@address:port/path)') },
+  ], [
+    { scm_url: 'file' },
+    {},
+  ], [
+    { scm_url: 'file:' },
+    {},
+  ], [
+    { scm_url: 'file:/' },
+    {},
+  ], [
+    { scm_url: 'file://' },
+    {},
+  ], [
+    { scm_url: '://acme.co' },
+    {},
+  ], [
+    { scm_url: 'x_test://acme.co' },
+    { scm_url: __('URL must include a protocol (http://, https:// or file://) or be a valid SSH path (user@server:path or ssh://user@address:port/path)') },
+  ], [
+    { scm_url: '-test://acme.co' },
+    { scm_url: __('URL must include a protocol (http://, https:// or file://) or be a valid SSH path (user@server:path or ssh://user@address:port/path)') },
+  ], [
+    { scm_url: '0test://acme.co' },
+    { scm_url: __('URL must include a protocol (http://, https:// or file://) or be a valid SSH path (user@server:path or ssh://user@address:port/path)') },
+  ], [
+    { scm_url: '+test://acme.co' },
+    { scm_url: __('URL must include a protocol (http://, https:// or file://) or be a valid SSH path (user@server:path or ssh://user@address:port/path)') },
+  ], [
+    { scm_url: '.test://acme.co' },
+    { scm_url: __('URL must include a protocol (http://, https:// or file://) or be a valid SSH path (user@server:path or ssh://user@address:port/path)') },
+  ], [
+    { scm_url: 'schem%6e://' },
+    { scm_url: __('URL must include a protocol (http://, https:// or file://) or be a valid SSH path (user@server:path or ssh://user@address:port/path)') },
+  ], [
+    { scm_url: 'x-Test+v1.0://acme.co' },
+    { scm_url: __('URL must include a protocol (http://, https:// or file://) or be a valid SSH path (user@server:path or ssh://user@address:port/path)') },
+  ],
+];
+
+// for fail caase
+// {scm_url = __('URL must include a protocol (http://, https:// or file://) or be a valid SSH path (user@server:path or ssh://user@address:port/path)')}
+
+describe('Helpers', () => {
+  it('urlValidation', () => {
+    for (const test of TESTCASES) {
+      expect(urlValidation(test[0])).toEqual(test[1]);
```

Comment:
to simplify, perhaps make the second value above a simple boolean, like `succeeds: true` or `succeeds: false`, then here do the following which will keep some of the copy/paste down to a minimum.

```suggestion
            if test[1]
              expectation = {}
            else
              expectation = { scm_url: __('URL must include a protocol (http://, https:// or file://) or be a valid SSH path (user@server:path or ssh://user@address:port/path)') }
            end
            expect(urlValidation(test[0])).toEqual(expectation);
```